### ISSUE TITLE
no-unused-vars false-positive with array reduce function

### ISSUE THREAD

**Issue by rugk** [eslint/eslint#13855]
[no-unused-vars](https://eslint.org/docs/rules/no-unused-vars) produces a false-positive if you make use of the Array reduce function and name a parameter the same as you name it inside of the lambda functions, apparently.

**Tell us about your environment**

* **ESLint Version:** 7.13.0
* **Node Version:** web browser
* **npm Version:** N/A

These questions make no sense, because I could reproduce it on https://eslint.org/demo, see below. (Also reproducible with EsLint v4.19.1 locally, however.)

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
the default

**Please show your full configuration:**

(nothing special, just the default)

<details>
<summary>Configuration</summary>

```js
{
    "parserOptions": {
        "ecmaVersion": 10,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "constructor-super": 2,
        "for-direction": 2,
        "getter-return": 2,
        "no-async-promise-executor": 2,
        "no-case-declarations": 2,
        "no-class-assign": 2,
        "no-compare-neg-zero": 2,
        "no-cond-assign": 2,
        "no-const-assign": 2,
        "no-constant-condition": 2,
        "no-control-regex": 2,
        "no-debugger": 2,
        "no-delete-var": 2,
        "no-dupe-args": 2,
        "no-dupe-class-members": 2,
        "no-dupe-else-if": 2,
        "no-dupe-keys": 2,
        "no-duplicate-case": 2,
        "no-empty": 2,
        "no-empty-character-class": 2,
        "no-empty-pattern": 2,
        "no-ex-assign": 2,
        "no-extra-boolean-cast": 2,
        "no-extra-semi": 2,
        "no-fallthrough": 2,
        "no-func-assign": 2,
        "no-global-assign": 2,
        "no-import-assign": 2,
        "no-inner-declarations": 2,
        "no-invalid-regexp": 2,
        "no-irregular-whitespace": 2,
        "no-misleading-character-class": 2,
        "no-mixed-spaces-and-tabs": 2,
        "no-new-symbol": 2,
        "no-obj-calls": 2,
        "no-octal": 2,
        "no-prototype-builtins": 2,
        "no-redeclare": 2,
        "no-regex-spaces": 2,
        "no-self-assign": 2,
        "no-setter-return": 2,
        "no-shadow-restricted-names": 2,
        "no-sparse-arrays": 2,
        "no-this-before-super": 2,
        "no-undef": 2,
        "no-unexpected-multiline": 2,
        "no-unreachable": 2,
        "no-unsafe-finally": 2,
        "no-unsafe-negation": 2,
        "no-unused-labels": 2,
        "no-unused-vars": 2,
        "no-useless-catch": 2,
        "no-useless-escape": 2,
        "no-with": 2,
        "require-yield": 2,
        "use-isnan": 2,
        "valid-typeof": 2
    },
    "env": {
        "browser": true,
        "node": true,
        "es6": true
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function getCssForMediaQueryFunc(queryString, stylesheet = document.styleSheets, previousText = "") {
    return Array.from(stylesheet).reduce((previousText, styleSheet) => {
        return Array.from(styleSheet.cssRules).reduce((previousText, cssRule) => {
            return previousText + cssRule;
        }, previousText);
    }, "");
}
getCssForMediaQueryFunc("dasdas");
```

(This example makes no sense. It was just adapted to show the error.)

**[click here for the web demo](https://eslint.org/demo#eyJ0ZXh0IjoiZnVuY3Rpb24gZ2V0Q3NzRm9yTWVkaWFRdWVyeUZ1bmMocXVlcnlTdHJpbmcsIHN0eWxlc2hlZXQgPSBkb2N1bWVudC5zdHlsZVNoZWV0cywgcHJldmlvdXNUZXh0ID0gXCJcIikge1xuICAgIHJldHVybiBBcnJheS5mcm9tKHN0eWxlc2hlZXQpLnJlZHVjZSgocHJldmlvdXNUZXh0LCBzdHlsZVNoZWV0KSA9PiB7XG4gICAgICAgIHJldHVybiBBcnJheS5mcm9tKHN0eWxlU2hlZXQuY3NzUnVsZXMpLnJlZHVjZSgocHJldmlvdXNUZXh0LCBjc3NSdWxlKSA9PiB7XG4gICAgICAgICAgICByZXR1cm4gcHJldmlvdXNUZXh0ICsgY3NzUnVsZTtcbiAgICAgICAgfSwgcHJldmlvdXNUZXh0KTtcbiAgICB9LCBcIlwiKTtcbn1cbmdldENzc0Zvck1lZGlhUXVlcnlGdW5jKFwiZGFzZGFzXCIpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTAsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2V0dGVyLXJldHVybiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7ImJyb3dzZXIiOnRydWUsIm5vZGUiOnRydWUsImVzNiI6dHJ1ZX19fQ==)**

**What did you expect to happen?**
No error should be reported.

**What actually happened? Please include the actual, raw output from ESLint.**
> 1:82 - 'previousText' is assigned a value but never used. ([no-unused-vars](https://eslint.org/docs/rules/no-unused-vars))

This is a bug, because actually it is of course used, just some nested lambda functions lower.
To do so, it is passed as the last argument to [Array.prototype.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce).

It can be shown that this is correct if you change the variable name `previousText` *only* in the argument part and at the end:
```diff
*** 1,8 ****
! function getCssForMediaQueryFunc(queryString, stylesheet = document.styleSheets, previousText = "") {
      return Array.from(stylesheet).reduce((previousText, styleSheet) => {
          return Array.from(styleSheet.cssRules).reduce((previousText, cssRule) => {
              return previousText + cssRule;
!         }, previousText);
      }, "");
  }
  getCssForMediaQueryFunc("dasdas");
--- 1,8 ----
! function getCssForMediaQueryFunc(queryString, stylesheet = document.styleSheets, newName = "") {
      return Array.from(stylesheet).reduce((previousText, styleSheet) => {
          return Array.from(styleSheet.cssRules).reduce((previousText, cssRule) => {
              return previousText + cssRule;
!         }, newName);
      }, "");
  }
  getCssForMediaQueryFunc("dasdas");
```

So [this changed version now works without problems](https://eslint.org/demo#eyJ0ZXh0IjoiZnVuY3Rpb24gZ2V0Q3NzRm9yTWVkaWFRdWVyeUZ1bmMocXVlcnlTdHJpbmcsIHN0eWxlc2hlZXQgPSBkb2N1bWVudC5zdHlsZVNoZWV0cywgbmV3TmFtZSA9IFwiXCIpIHtcbiAgICByZXR1cm4gQXJyYXkuZnJvbShzdHlsZXNoZWV0KS5yZWR1Y2UoKHByZXZpb3VzVGV4dCwgc3R5bGVTaGVldCkgPT4ge1xuICAgICAgICByZXR1cm4gQXJyYXkuZnJvbShzdHlsZVNoZWV0LmNzc1J1bGVzKS5yZWR1Y2UoKHByZXZpb3VzVGV4dCwgY3NzUnVsZSkgPT4ge1xuICAgICAgICAgICAgcmV0dXJuIHByZXZpb3VzVGV4dCArIGNzc1J1bGU7XG4gICAgICAgIH0sIG5ld05hbWUpO1xuICAgIH0sIFwiXCIpO1xufVxuZ2V0Q3NzRm9yTWVkaWFRdWVyeUZ1bmMoXCJkYXNkYXNcIik7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMCwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnsiYnJvd3NlciI6dHJ1ZSwibm9kZSI6dHJ1ZSwiZXM2Ijp0cnVlfX19).

**Are you willing to submit a pull request to fix this bug?**
No

**Comment by ljharb** [comment]
Your change is actually a bugfix. In your example code, the `previousText` passed into the outer reduce is not used; but instead the inner reduce is using the accumulator.

In other words, the correct code is this:
```diff
function getCssForMediaQueryFunc(queryString, stylesheet = document.styleSheets, previousText = "") {
    return Array.from(stylesheet).reduce((previousText, styleSheet) => {
        return Array.from(styleSheet.cssRules).reduce((previousText, cssRule) => {
            return previousText + cssRule;
-        }, "");
+        }, previousText);
    }, previousText);
}
getCssForMediaQueryFunc("dasdas");
```

**Comment by rugk** [comment]
Ahhhh, :facepalm: of course!

That's the nestyness of nested functions, yeah, of course. Thanks! :smiley:

**Comment by ljharb** [comment]
np. i suggest you enable the no-shadow rule, so you're forced to rename these variables, and you'll more easily spot the issues :-)
